Add Progressively Enhance your PWA article

tomayac opened the pull request:
Fixes #1442 

Changes proposed in this pull request:

- Add the article.

jpmedley commented:
@kaycebasques I'm reviewing this, FYI.

tomayac commented:
> @kaycebasques I'm reviewing this, FYI.

@jpmedley Thanks for the review offer! Just note that it's not ready for review quite yet.

jpmedley commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,831 @@
+---
+
+Back in March 2003, Nick Finck and Steve Champeon stunned the web design world
+with the concept of progressive enhancement.
+A strategy for web design that emphasizes core webpage content first,
```
```suggestion
with the concept of progressive enhancement,
a strategy for web design that emphasizes core webpage content first,
```

tomayac replied:
Done

jpmedley commented:
>@jpmedley Thanks for the review offer! Just note that it's not ready for review quite yet.

I just figured that out.

kaycebasques commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+As an installable PWA, Fugu Greetings of course does have an app icon
+that users can place on the app dock or the home screen.
+A fun and easy way to demonstrate the API is to (ab)use it in Fugu Greetings
+is to use it as a pen strokes counter.
```
```suggestion
as a pen strokes counter.
```

kaycebasques commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+## The sample app: Fugu Greetings
+
+For this article, I work with a simple PWA, called
+[Fugu Greetings](https://tomayac.github.io/fugu-greetings/public/).
```
Also link to the repository?

tomayac replied:
Good idea.

kaycebasques commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+This applies to both the client code and the service worker code.
+On non-supporting browsers, neither of them is loaded.
+Note how in the service worker, instead of a dynamic `import()`, I use the classic
+[`importScripts()`](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts).
```
Why `importScripts()` instead of `import()`?

tomayac replied:
Because in service workers `import()` isn't supported yet. I'll add that info.

kaycebasques commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+
+I also want to include the [Wake Lock API](https://web.dev/wakelock/).
+Sometimes you just need to stare long enough at the screen until inspiration
+kisses you.
```
> until inspiration kisses you

lol great phrase

tomayac replied:
💋💡

kaycebasques commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+
+## Closing
+
+Phew, what a ride. So many APIs in just one sample app.
```
A long journey, but informative!

tomayac replied:
Thank you :-) The idea is to show as many APIs in as non-contrived context as possible.

kaycebasques commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+<figure class="w-figure">
+  <img class="w-screenshot"
+       src="10000201000009C4000006E8D776365D6B538C9A.png"
+       alt="Chrome DevTools Network tab showing only requests for files with code that the current browser supports.">
```
```suggestion
       alt="Chrome DevTools Network panel showing only requests for files with code that the current browser supports.">
```
In the official docs I always refer to the top-level UI containers as `panels`. I believe the phrase predated me. I know `tab` is a more intuitive term but it would be nice to keep the alignment with the official docs terminology throughout this post.

kaycebasques commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+  </figcaption>
+</figure>
+
+We are working hard on making the grass greener when it comes to advanced Fugu APIs.
```
```suggestion
Chrome is working hard on making the grass greener when it comes to advanced Fugu APIs.
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+with the concept of
+[progressive enhancement](http://www.hesketh.com/publications/inclusive_web_design_for_the_future/),
+a strategy for web design that emphasizes core webpage content first,
+and that then progressively adds more nuanced
```
"adds" is correct here, the strategy adds.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+
+<figure class="w-figure w-figure--fullbleed">
+  <img src="1000020100000640000003E810E16D93C747B2D5.png"
+       alt="The iconic Windows XP green grass background image.">
```
This was in the original talk. It's more of a pun.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+## The sample app: Fugu Greetings
+
+For this article, I work with a simple PWA, called
+[Fugu Greetings](https://tomayac.github.io/fugu-greetings/public/).
```
```suggestion
[Fugu Greetings](https://tomayac.github.io/fugu-greetings/public/)
([GitHub](https://github.com/tomayac/fugu-greetings)).
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1052 @@
+
+<blockquote>
+  <p>
+    Progressive enhancement is a design philosophy that centers around providing a baseline of
+    essential content and functionality to as many users as possible, while at the same time going
+    further and delivering the best possible experience only to users of the most modern browsers
+    that can run all the required code.
+  </p>
+  <p>
+    Feature detection is generally used to determine whether browsers can handle the high level
+    content or not, with polyfills often being used to build in missing features with JavaScript.
+  </p>
+  <p>[…]</p>
+  <p>
+    It is a useful technique that allows Web developers to focus on developing the best possible
+    websites while balancing the issues in those websites being accessed by multiple unknown
+    user-agents. Graceful degradation is related but different—often seen as going in the opposite
+    direction to progressive enhancement. In reality both approaches are valid and can often
+    complement one another.
```
```suggestion
    Progressive enhancement is a design philosophy that povides a baseline
    of essential content and functionality to as many users as possible,
    while delivering the best possible experience only to users
    of the most modern browsers that can run all the required code.
  </p>
  <p>
    [Feature detection](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Feature_detection)
    is generally used to determine whether browsers can handle more modern functionality,
    while [polyfills](https://developer.mozilla.org/en-US/docs/Glossary/Polyfill)
    are often used to add missing features with JavaScript.
  </p>
  <p>[…]</p>
  <p>  
    Progressive enhancement is a useful technique that allows web developers to focus
    on developing the best possible websites while making those websites work
    on multiple unknown user agents.
    [Graceful degradation](https://developer.mozilla.org/en-US/docs/Glossary/Graceful_degradation)
    is related, but is not the same thing and is often seen as going in the opposite direction
    to progressive enhancement.
    In reality, both approaches are valid and can often complement one another.
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+    direction to progressive enhancement. In reality both approaches are valid and can often
+    complement one another.
+  </p>
+</blockquote>
```
Perfect, took the new MDN text now.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+</blockquote>
+
+{%Aside 'note' %}
+  This is not an article on progressive enhancement, but assumes you are familiar
```
I just mean this is not an intro to the concept of PE per se.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1052 @@
+</blockquote>
+
+{%Aside 'note' %}
+  This is not an article on progressive enhancement, but assumes you are familiar
```
```suggestion
  This is not an introductory article on progressive enhancement, but assumes you are familiar
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+[`<input type=file>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file)
+element to make this happen.
+First, you'd create the element, set its `type` and the to-be-`accept`ed MIME types,
+and then programmatically "click" it and listen for changes.
```
No, you just call `.click()`, it's a programmatic click.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1054 @@
+```
+
+As before, I use a progressive enhancement loading strategy.
+If both `'share'` and `'canShare'` exist on the `navigator` object, only then I go forward and
````
Which we do, in this case we're really looking for the strings.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1052 @@
+or contacts manager app.
+When you write a greeting card, it may not always be easy to correctly write
+someone's name.
+For example, I have a friend who prefers their name to be spelled in Cyrillic letters. I'm
```
```suggestion
For example, I have a friend Sergey who prefers his name to be spelled in Cyrillic letters. I'm
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+I mean contacts as in a device's address book or contacts manager app.
+When you write a greeting card, it may not always be easy to correctly write
+someone's name.
+For example, I have a friend who prefers their name to be spelled in Cyrillic letters. I'm
```
It's a pun since my friends are Larry and Sergey, but I have added Sergey now to use his.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1051 @@
+}
+```
+
+In Fugu Greeting, when I tap the Contacts button and select my two best pals,
+[Сергей Михайлович Брин](https://ru.wikipedia.org/wiki/%D0%91%D1%80%D0%B8%D0%BD,_%D0%A1%D0%B5%D1%80%D0%B3%D0%B5%D0%B9_%D0%9C%D0%B8%D1%85%D0%B0%D0%B9%D0%BB%D0%BE%D0%B2%D0%B8%D1%87) and [劳伦斯·爱德华·"拉里"·佩奇](https://zh.wikipedia.org/wiki/%E6%8B%89%E9%87%8C%C2%B7%E4%BD%A9%E5%A5%87),
````
🤡

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1052 @@
+A fun and easy way to demonstrate the API is to (ab)use it in Fugu Greetings
+is to use it as a pen strokes counter.
+I have added an event listener that on the `pointerdown` event increments the pen strokes
+counter and sets the icon.
```
```suggestion
and then sets the updated icon badge.
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1052 @@
+API is supported by the browser.
+This applies to both the client code and the service worker code.
+On non-supporting browsers, neither of them is loaded.
+Note how in the service worker, instead of a dynamic `import()`, I use the classic
```
```suggestion
Note how in the service worker, instead of a dynamic `import()`
(that isn't supported in a service worker context yet), I use the classic
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+  </figcaption>
+</figure>
+
+## Notification Triggers API
```
I don't want this article to be a maintenance nightmare. The individual articles like https://web.dev/notification-triggers/ will be updated with the current status.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1052 @@
+## The Idle Detection API
+
+At times, even if you stare at the screen for hours,
+it's just useless.
```
```suggestion
it's just useless and you can't come up with any idea what to do with your greeting card.
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+
+## The Idle Detection API
+
+At times, even if you stare at the screen for hours,
+it's just useless.
```
I added more context.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+and clears the canvas.
+This API is currently gated behind the
+[notifications permission](https://developer.mozilla.org/en-US/docs/Web/API/Notification/requestPermission),
+since a lot of production use cases of idle detection are notifications-related,
+for example, to only send a notification to a device the user is currently actively using.
```
It doesn't. Which is part of the discussion in https://github.com/WICG/idle-detection/issues/21.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+This applies to both the client code and the service worker code.
+On non-supporting browsers, neither of them is loaded.
+Note how in the service worker, instead of a dynamic `import()`
+(that isn't supported in a service worker context yet), I use the classic
```
```suggestion
(that isn't supported in a service worker context
[yet](https://github.com/w3c/ServiceWorker/issues/1356#issuecomment-433411852)),
I use the classic
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+
+I'm grateful to [Christian Liebel](https://christianliebel.com/) and
+[Hemanth HM](https://h3manth.com/) who both have contributed to Fugu Greetings.
+This article was reviewed by [Joe Medley](https://github.com/jpmedley).
```
```suggestion
This article was reviewed by [Joe Medley](https://github.com/jpmedley) and
[Kayce Basques](https://github.com/kaycebasques).
[Jake Archibald](https://github.com/jakearchibald/) helped me find out the situation
with dynamic `import()` in a service worker context.
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+
+The code to make this happen is pretty straightforward. I call `navigator.share()` and
+pass it an optional `title`, `text`, and `url` in an object.
+But what if I want to attach an image? Level&nbsp;1 of the Web Share API doesn't support this yet.
+The good news is that Web Share Level&nbsp;2 has added file sharing capabilities.
```
Here, I respectfully disagree. This is meant to show the evolvement from v1 to v2. Simple to more complex.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1054 @@
+};
+```
+
+The second step is to *listen* for the periodic sync event in the service worker.
````
Before I emphasize register, now listen. It's two steps.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+    Chrome DevTools network tab.
+  </figcaption>
+</figure>
+
```
Thank you :-)

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1051 @@
+```js
+try {
+  await navigator.share({
+    title: '',
````
The title doesn't really do much (it's mostly ignored in current implementations). I'll add one, though.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1059 @@
+```js
+try {
+  await navigator.share({
+    title: '',
````
```suggestion
    title: 'Check out this article:',
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1059 @@
+<figure class="w-figure">
+  <img class="w-screenshot"
+       src="10000201000006FC0000021AD16A28BA8F5CBF1A.png"
+       alt="">
```
```suggestion
       alt="Chrome DevTools showing the modern files getting loaded.">
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1059 @@
+};
+```
+
+In Fugu Greetings, if I tap the Share button on a supporting browser like Chrome on Android,
````
```suggestion
In Fugu Greetings, if I tap the **Share** button on a supporting browser like Chrome on Android,
```

tomayac replied:
```suggestion
In Fugu Greetings, if I tap the Share button on a supporting browser like Chrome on Android,
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1059 @@
+mechanisms.
+For example, below is desktop Safari's share sheet on macOS triggered from an article on
+my [blog](https://blog.tomayac.com/).
+When you click the Share Article button, you can share a link to the article with a friend, for
```
```suggestion
When you click the **Share Article** button, you can share a link to the article with a friend, for
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1059 @@
+}
+```
+
+In Fugu Greeting, when I tap the Contacts button and select my two best pals,
````
```suggestion
In Fugu Greeting, when I tap the **Contacts** button and select my two best pals,
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1059 @@
+}
+```
+
+In Fugu Greetings, pressing the Wallpaper button reveals the greeting card image of the day
````
```suggestion
In Fugu Greetings, pressing the **Wallpaper** button reveals the greeting card image of the day
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1059 @@
+       src="10000201000009C4000005E0B042853D0E77C224.png"
+       alt="Fugu Greetings app with a new greeting card image of the day.">
+  <figcaption class="w-figcaption">
+    Pressing the Wallpaper button displays the image of the day.
```
```suggestion
    Pressing the <strong>Wallpaper</strong> button displays the image of the day.
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1059 @@
+
+So how does this work in practice? I have an image open in the macOS Preview app and
+copy it to the clipboard.
+When I click Paste, the Fugu Greetings app then asks me
```
```suggestion
When I click **Paste**, the Fugu Greetings app then asks me
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1059 @@
+Finally, after accepting the permission, the image is then pasted into the application.
+The other way round works, too.
+Let me copy a greeting card to the clipboard.
+When I then open Preview and click "File" and then "New from Clipboard",
```
```suggestion
When I then open Preview and click **File** and then **New from Clipboard**,
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1059 @@
+  import('./wake_lock.mjs');
+}
+```
+In Fugu Greetings, there's an insomnia checkbox that, when checked, keeps the
````
```suggestion
In Fugu Greetings, there's an **Insomnia** checkbox that, when checked, keeps the
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1059 @@
+       src="10000201000009C4000005DBCD9D10DD0A745315.png"
+       alt="The insomnia checkbox, if checked, keeps the screen awake.">
+  <figcaption class="w-figcaption">
+    The insomnia checkbox keeps app awake.
```
```suggestion
    The <strong>Insomnia</strong> checkbox keeps app awake.
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1059 @@
+}
+```
+
+In the Fugu Greetings app, the canvas clears when the Ephemeral checkbox is
````
```suggestion
In the Fugu Greetings app, the canvas clears when the **Ephemeral** checkbox is
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1059 @@
+       src="10000201000009C4000005DBB930B1D3F8F5D1DA.png"
+       alt="Fugu Greetings app with a cleared canvas after the user has been idle for too long.">
+  <figcaption class="w-figcaption">
+    When the ephemeral checkbox is checked and the user has been idle for too long, the canvas is cleared.
```
```suggestion
    When the <strong>Ephemeral</strong> checkbox is checked and the user has been idle for too long, the canvas is cleared.
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
````diff
@@ -0,0 +1,1059 @@
+}
+```
+
+When I check the Reminder checkbox in Fugu Greetings, a prompt asks
````
```suggestion
When I check the **Reminder** checkbox in Fugu Greetings, a prompt asks
```

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1054 @@
+<figure class="w-figure">
+  <img class="w-screenshot"
+       src="10000201000006FC0000021AD16A28BA8F5CBF1A.png"
+       alt="">
```
Good find. Fixed.

tomayac commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1051 @@
+Finally, after accepting the permission, the image is then pasted into the application.
+The other way round works, too.
+Let me copy a greeting card to the clipboard.
+When I then open Preview and click "File" and then "New from Clipboard",
```
Fixed.

tomayac commented:
I’ve addressed all the review comments, I think it’s ready from an editorial point of view. Pending: image optimization. Please not that this should go live with or after web.dev Live, not earlier (hence the Do Not Merge label. Thanks for the feedback!

kaycebasques commented on src/site/content/en/blog/progressively-enhance-your-pwa/index.md:
```diff
@@ -0,0 +1,1059 @@
+  new web capabilities like native file system access, system clipboard access,
+  contacts retrieval, periodic background sync, screen wake lock, web sharing features,
+  and many more.
+date: 2020-06-12
+updated: 2020-06-15
```
```suggestion
scheduled: true
date: 2020-07-01
# updated: 2020-06-15
```
@tomayac let's just merge this and rely on the fancy `scheduled: true` feature to deploy this at 7:00AM (PST) on the day of your talk?

https://web.dev/handbook/markup-post-codelab/#set-up-the-yaml

tomayac commented:
@kaycebasques All images were shrunk in https://github.com/GoogleChrome/web.dev/pull/3178/commits/a3fd45eee585c9b1fcf98043ca971bb8c9527110, using ImageOptim lossy compression with 70% quality. PTAL.

I have committed your suggested change so the post is scheduled to go live during web.dev Live. Thanks!
